File: day17.py
"""Solution to day 17 of Advent of Code"""

from get_input import get_input
from collections import defaultdict
import itertools


def print_grid(grid):
    for z in range(-2, 3, 1):
        print(f"z={z}")
        for r in range(-1, 4):
            for c in range(-1, 4):
                print(grid.get((r, c, z), '.'), end='')
            print()
        print()
    print('-'*10)


def neighbors(point):
    for diff in itertools.product((-1, 0, 1), repeat=len(point)):
        if all(d == 0 for d in diff):
            continue
        neighbor = tuple(a + b for a, b in zip(point, diff))
        yield neighbor


def part1(grid):
    current = {k: v for k, v in grid.items() if v == '#'}
    for _ in range(6):
        prev = current
        current = {}
        active = set(n for p, v in prev.items()
                     for n in neighbors(p) if v == '#')
        for point in active:
            value = prev.get(point, '.')
            count = sum(1 for n in neighbors(point) if prev.get(n, '.') == '#')
            if value == '#' and (count == 2 or count == 3):
                current[point] = '#'
            elif value == '.' and count == 3:
                current[point] = '#'
    return len(current)


def part2(grid):
    current = {}
    for k, v in grid.items():
        k = k + (0,)
        if v != '#':
            continue
        current[k] = v
    for _ in range(6):
        prev = current
        current = {}
        active = set(n for p, v in prev.items()
                     for n in neighbors(p) if v == '#')
        for point in active:
            value = prev.get(point, '.')
            count = sum(1 for n in neighbors(point) if prev.get(n, '.') == '#')
            if value == '#' and (count == 2 or count == 3):
                current[point] = '#'
            elif value == '.' and count == 3:
                current[point] = '#'
    return len(current)


def parse(text):
    result = defaultdict(lambda: '.')
    for r, row in enumerate(text.splitlines()):
        row = row.strip()
        for c, value in enumerate(row):
            result[(r, c, 0)] = value
    return result


TEST = """.#.
..#
###"""


if __name__ == "__main__":
    assert part1(parse(TEST)) == 112
    assert part2(parse(TEST)) == 848
    LINES = parse(get_input(day=17, year=2020))
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day17.rs
use common::get_input;
use std::collections::{HashMap, HashSet};
use itertools::Itertools;

type Point = Vec<isize>;
type Input = HashMap<Point, char>;

fn neighbors(p: &Point) -> Vec<Point> {
    (0..p.len()).map(|_| -1..2).multi_cartesian_product()
        .filter_map(|d|
            if d.iter().all(|v| v == &0) {
                None
            } else {
                Some(p.iter().zip(d.iter()).map(|(i, j)| i + j).collect())
            })
        .collect()
}

fn part1(input: &Input) -> usize {
    let mut current: HashMap<Point, char> = input.iter()
        .filter_map(|(k, v)| if v == &'#' { Some((k.clone(), *v)) } else { None })
        .collect();
    for _ in 0..6 {
        let prev = current;
        current = HashMap::new();
        let active: HashSet<Point> = prev.iter()
            .filter_map(|(k, v)| if v == &'#' { Some(neighbors(k)) } else { None }).flatten().collect();
        for point in active {
            let value = prev.get(&point).unwrap_or(&'.');
            let count = neighbors(&point).iter().filter(|&p| Some(&'#') == prev.get(p)).count();
            if vec![(2, '#'), (3, '#'), (3, '.')].contains(&(count, *value)) {
                current.insert(point, '#');
            }
        }
    }
    current.len()
}

fn part2(input: &Input) -> usize {
    let mut current: HashMap<Point, char> = HashMap::new();
    for (k, v) in input.iter() {
        if v != &'#' {
            continue
        }
        let mut key = k.clone();
        key.push(0);
        current.insert(key, '#');
    }
    for _ in 0..6 {
        let prev = current;
        current = HashMap::new();
        let active: HashSet<Point> = prev.iter()
            .filter_map(|(k, v)| if v == &'#' { Some(neighbors(k)) } else { None }).flatten().collect();
        for point in active {
            let value = prev.get(&point).unwrap_or(&'.');
            let count = neighbors(&point).iter().filter(|&p| Some(&'#') == prev.get(p)).count();
            if vec![(2, '#'), (3, '#'), (3, '.')].contains(&(count, *value)) {
                current.insert(point, '#');
            }
        }
    }
    current.len()
}

fn parse(text: String) -> Input {
    text.trim().split('\n').enumerate().map(|(r, row)| {
        row.chars().enumerate().map(move |(c, chr)| (vec![c as isize, r as isize, 0], chr))
    }).flatten().collect()
}

fn main() {
    assert_eq!(part1(&parse(".#.\n..#\n###".to_string())), 112);
    assert_eq!(part2(&parse(".#.\n..#\n###".to_string())), 848);
    let input = parse(get_input(17, 2020));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
